src/main/scala/scala/collection/next/NextIterableOnceOpsExtensions.scala

```diff
@@ -0,0 +1,101 @@
+/*
+ * Scala (https://www.scala-lang.org)
+ *
+ * Copyright EPFL and Lightbend, Inc.
+ *
+ * Licensed under Apache License 2.0
+ * (http://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package scala.collection
+package next
+
+import scala.language.implicitConversions
+
+private[next] final class NextIterableOnceOpsExtensions[A, CC[_], C](
+  private val col: IterableOnceOps[A, CC, C]
+) extends AnyVal {
+  import NextIterableOnceOpsExtensions.GroupMap
+
+  /**
+   * Partitions this IterableOnce into a map according to a discriminator function `key`. All the values that
+   * have the same discriminator are then transformed by the `value` function and then reduced into a
+   * single value with the `reduce` function.
+   *
+   * {{{
+   *   def occurrences[A](as: IterableOnce[A]): Map[A, Int] =
+   *     as.iterator.groupMapReduce(identity)(_ => 1)(_ + _)
+   * }}}
+   *
+   * @note This will force the evaluation of the Iterator.
+   */
+  def groupMapReduce[K, B](key: A => K)(f: A => B)(reduce: (B, B) => B): immutable.Map[K, B] =
+    groupMapTo(key)(f).reduce(reduce)
+
+  def groupByTo[K](key: A => K): GroupMap[A, K, A, immutable.Iterable, immutable.Map] =
+    groupMapTo(key)(identity)
+
+  def groupMapTo[K, V](key: A => K)(f: A => V): GroupMap[A, K, V, immutable.Iterable, immutable.Map] =
+    new GroupMap(col, key, f, immutable.Iterable, immutable.Map)
+}
+
+object NextIterableOnceOpsExtensions {
+  final case class GroupMap[A, K, V, CC[_], MC[_, _]](
+    col: IterableOnceOps[A, Any, Any],
+    key: A => K,
+    f: A => V,
+    colFactory: Factory[V, CC[V]],
+    mapFactory: CustomMapFactory[MC, K]
```

Comment:
any reason not to just use a `Factory[(K, V), M]` (where `M` is not parameterised/not a type constructor)

Reply:
The reason for **CustomMapFactory** is not really the `MC[_, _]` is the fact that for **SortedMaps** it needs to know that there is an **Ordering** for `K`.

Now, I have to be honest that I tried so many things like having `map` inside **GroupMap** _(but that opened the door to bugs)_ that maybe with the current design it shouldn't be needed anymore.

However, the other reason was to allow changing the `V` since if you call `result` it consumes `(K, C)` whereas if you call `reduce` it consumes `(K, V)`. Which, I am pretty sure is not fixable right now. Or maybe with existential?
Other option would be to move `reduce` to other API, which would have the unnecessary `collectValuesAs`.